### ISSUE TITLE
A way to explicitly tell which source fields to null check

### ISSUE THREAD

**Issue by PCasafont** [mapstruct/mapstruct#1243]
Currently, when we create an update mapping such as:
`User updateUserFromDto(@MappingTarget User user, UserDto userDto);`

In order to have null checks, we must configure the Mapper annotation (in the class level) to have nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS.

But this generates null checks everywhere, including the creator mappers.
It would be awesome if there was a property in @Mapping like "nullCheck=true/false", so that we can fully control that.

**Comment by filiphr** [comment]
Doing it on the `@Mapping` is one way to do it.

I think that it could be better if we can check for `NotNull` or `Nullable` annotations on the getters. This is what IntelliJ [does](https://www.jetbrains.com/help/idea/nullable-and-notnull-annotations.html) for the code inspections. I also think that Kotlin uses those annotations as for interoperability with Java.  Spring has also done it for their entire code base, see [SPR-15540](https://jira.spring.io/browse/SPR-15540). I am not sure about other frameworks. 

WDYT?

**Comment by PCasafont** [comment]
A check for nullable/notnull annotations would be awesome!

El dia 16/07/2017 13:39, "Filip Hrisafov" <notifications@github.com> va
escriure:

Doing it on the @Mapping is one way to do it.

I think that it could be better if we can check for NotNull or Nullable
annotations on the getters. This is what IntelliJ does
<https://www.jetbrains.com/help/idea/nullable-and-notnull-annotations.html>
for the code inspections. I also think that Kotlin uses those annotations
as for interoperability with Java. Spring has also done it for their entire
code base, see SPR-15540 <https://jira.spring.io/browse/SPR-15540>. I am
not sure about other frameworks.

WDYT?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/mapstruct/mapstruct/issues/1243#issuecomment-315603411>,
or mute the thread

**Comment by brunomarquete** [comment]
Are there plans to start this feature? I am working with non-null objects in Kotlin and I would like to see this feature implemented in MapStruct.

**Comment by altarasyuk** [comment]
It would be good feature

**Comment by mdelmoral** [comment]
In my current company we use the library in small and not very critical services since we are very strict with the validation of nullity. This feature would be fundamental to adopt the library at a global level 👍🏻

**Comment by darioseidl** [comment]
For Java, JSR 303 support would be nice. Similarly for Kotlin, it would be great if the implementation could create null-checks depending on the nullability of the Kotlin types in the mapper interface.

For example, in the [PersonConverter](https://github.com/mapstruct/mapstruct-examples/blob/master/mapstruct-kotlin-gradle/src/main/kotlin/org/mapstruct/example/kotlin/converter/PersonConverter.kt) all types are non-nullable, but in the generated `PersonConverterImpl`, null would be passed through instead of throwing an exception.

**Comment by andye2004** [comment]
This would be an incredibly valuable feature and something that would encourage greater adoption of MapStruct. Are there any plans to start this work given the ticket has been open for close to four years now?

**Comment by anandbhaskaran** [comment]
This would be incredibly nice to have a feature!

**Comment by filiphr** [comment]
Thanks for your examples @agentgt.

>MapStruct should know to call that method to convert the nullable string type to a nonnull (or in this case stripped annotation) string type. After all it is a different type.

I am not sure that MapStruct should treat methods with types annotated differently as different types. I would rather have MapStruct properly detect `NonNull` / `Nullable` annotations and invoke the methods appropriately. 

Regarding the code that we should generate, I don't see a reason why not to generate code with passing down the annotations on the return type itself. However, that topic is different than this issue and I would suggest creating a separate issue where we can discuss MapStruct generating such code (irregardless of the bugs in the compilers)

**Comment by floydkretschmar** [comment]
Hello everyone, when evaluating in our team whether or not we should adapt mapstruct in our codebase, we ran into the exact same problem described by this issue. Namely a bunch of unnecessary null checks for code, that is already null safe.

I had a look at some of the proposals made in this issue and I also spend a few hours looking through the code, to see how much work it would be to implement `NullValueCheckStrategy.NEVER` as proposed by @agentgt. Meaning introducing a NullValueCheckStrategy where no null checks are done at all.

But it seems to me from how I understand the code, that doing null checks for right hand side values of an assignment is kind of core design assumption. This is emphasized by the fact that this issue has been open for a while now. I am basically wondering if it is worth taking a stab at implementing this new NullValueCheckStrategy as a contrubution.

**Comment by filiphr** [comment]
@floydkretschmar I don't think that the path with `NullValueCheckStrategy.NEVER` is the right path.

I think that the right solution would be 

> I would rather have MapStruct properly detect `NonNull` / `Nullable` annotations and invoke the methods appropriately.

Which means that if you have something like:

```java
interface Customer {

    @NonNull
    String getId();

    String getName();
}

class CustomerDto {

    public void setId(@NonNull String id) {}

    public void setName(@NonNull String name) {}

}

@Mapper
public interface CustomerMapper {

    CustomerDto map(@NonNull Customer customer);

}
```

We would generate something like:

```java
public class CustomerMapperImpl implements Customer {

    @Override
    public CustomerDto map(Customer customer) {
        CustomerDto dto = new CustomerDto();

        dto.setId( customer.getId() );

        if ( customer.getName() != null ) {
            dto.setName( customer.getName() );
        }

        return dto;
    }
}
```

You can see that the `null` check for `customer` is not there. We are also doing a null check for `customer.getName()` because `setName` has `@NonNull` and `getName` could be `null`. By default we are setting it directly.

---

The reason why this has been open for a while is because it goes into the fundamentals of MapStruct and it needs some good thinking and typing it out for how to do it properly. We currently do not have any mechanism to scan types and apply the JSR 303 annotations.

Since this is one of the most requested issues, I think that we can try to tackle it for the next release (after 1.6)

**Comment by floydkretschmar** [comment]
@filiphr Thanks for the answer. Absolutely agreed that conforming with JSR 303 is the best solution, I was just wondering if that was even on the roadmap still. But if it is, then thats great news!

**Comment by filiphr** [comment]
@agentgt thanks for pointing that out. My plan was indeed to also look into JSpecify when starting to work on this.

**Comment by agentgt** [comment]
@filiphr As a sort of meta tangental goal MapStruct code base could be annotated and that may provide education by doing on JSpecify as well as improving code base. Just an idea.

**Comment by filiphr** [comment]
Something certainly worth exploring @agentgt. It is indeed easier to annotate our code base that to make it work in the code generation bits 😄. 

I had a brief look at JSpecify and from the docs it doesn't look too complicated to use it. It basically has 4 annotations and it is only around nullness from what I could see.

**Comment by AaronNeufeld** [comment]
I know it's been a while since this has been active, but if I can add my two cents, can we please ensure that there is support (i.e. compiler option) to switch between "assume null" and "assume non-null" modes for non-annotated types? I'm working with [Checker Framework](https://checkerframework.org/manual/#nullness-checker) which starts from the assumption that everything is non-null unless annotated otherwise, so having to annotate as non-null to interop with MapStruct would be a non-starter. We already encounter this issue with other specs/libraries that assume null (i.e. JSR-380 Bean Validation, JPA), and it's a significant pain point. If MapStruct can incorporate this flexibility from the beginning, that would be amazing. Thanks.

**Comment by vmckinney-cainc** [comment]
Is there any update to this, or at least a work around?

**Comment by xenoterracide** [comment]
+1 . If I had a first choice it would be to detect `@Nullmarked` and simply throw an exception (personal Ideal `IllegalArgumentException`) if you pass a null and prevent the return of the null. I don't know/feel that the internals of mapstruct need to concern itself initially with this too much. Alternatively, if you said it would do that if I put `@NonNull` on the input and output I'd be ok with that too.

**Comment by xenoterracide** [comment]
as mentioned in another ticket currently jspecify can't work with `@AnnotatedWith` reason being

```java
@AnnotatedWith(Nullable.class)
Foo mapper( Bar bar )
```

Will throw an error that Nullable can't be applied to a method. That's because it's a type annotation, and generally speaking all nullity annotations should be type annotations because they are also meant to be able to do 

```java
List<@Nullable Foo> foos( List<@Nullable> bars );
```

With Spring Boot 4 stack mass adopting jspecify, one option might be to detect `@Nullmarked` and then mapstruct should just auto apply them to any generated code that it can take/return nullable (which is currently everything). Another would be to simply let type or nullity annotations be special and simply copy them (beware jakarta validation, it is a runtime annotation and not a nullity one).

The current wokaround is to put

```java
@AnnotatedWith(NullUnmarked.class)
```

on your interface/abstract class, which remove null checking inside of the implementation.

**Comment by chehsunliu** [comment]
My workaround is to make `errorprone` ignore the whole generated folder:

```kotlin
import net.ltgt.gradle.errorprone.CheckSeverity
import net.ltgt.gradle.errorprone.errorprone

tasks.withType<JavaCompile> {
    options.errorprone {
        excludedPaths.set(".*/build/generated/.*")

        check("NullAway", CheckSeverity.ERROR)
        option("NullAway:AnnotatedPackages", "io.github.xxx.ooo")
    }
}
```

I used to set up NullAway option for excluding classes, but it's too tedious.

**Comment by xenoterracide** [comment]
Sure, I wouldn't recommend ignoring entire packages. What we're doing is we're just adding the nullunmarked annotation to the generated class with annotatedwith But it would be much better if the implementation could just use the annotations that are added to the interface /package and implement correctly according to those annotations

**Comment by nealeu** [comment]
I'm currently tracking one of those painful null situations where data goes via MongoDB, so the value that should not be null in the database is passed into a JSpecify `@NullMarked` class and propagates through a lot of NullAway-backed code until ... NPE.
What would be useful would be for dev and QA environments running with `-enableassertions` to catch this by having non-null mappings also `assert value != null;`.  

This would seem to be a good option to include when mapping to a value expected to be non-null.

**Comment by nealeu** [comment]
We're getting some runtime benefits from something like the above approach with some Lombok support for @NullMarked which is adding configurable null checks to constructors and setters.  Our `NullSafety.assumeNonNull(Object x, String msg)` implementation which our Lombok fork uses to check nulls, allows us to throw NPE during test runs, but `log.error()` at runtime so we can catch false negatives.
However, we ultimately want to be able to prevent a copy of `@Nullable` to `@NonNull` via mapstruct generated code.  This could be an optional compile time check or warning for MapStruct, just as NullAway gives us the option to warn or error.

**Comment by pkernevez** [comment]
It would be nice that if we define a mapping like:
```
public Dto from(Entity e);
```

MapStruct is able to create a private mapping method and use it when a nullable type is detected somewhere (aka a getter returning `@Nullable Entity` )
```
private @Nullable Dto fromNullable(@Nullable Entity e){
if (e==null){
   return null;
} else { // Delegates to non null version
   return from(e);
}
```

**Comment by nealeu** [comment]
@filiphr I notice this is for discussion.
My take on where the Java language is going is that [explicit nullability](https://openjdk.org/jeps/8303099) will happen because it is a requirement for [flattening object storage in value types](https://openjdk.org/jeps/8316779).

That means that annotations that target parameters, fields and return values need to be phased out in favour of type use annotations for `@Nullable` and `@NonNull`, such as JSpecify which aligns with the general direction and, of course, Kotlin which already supports strict nullability.

I think this is such a big change that MapStruct (and Lombok too) will benefit from a longer term plan, which I think starts with Kotlin and JSpecify support.

An example use case would be to have MapStruct refuse to generate code that populates a non-null field from a nullable field without a default value.

This would mean that for the following code, MapStruct would issue an error rather than generate code that potentially writes `null` to a non-null field.

```java
@NullMarked
record Source(
  String firstName,
  @Nullable lastName
) {}

@NullMarked
record Target(
  String firstName,
  String lastName
) {}

@Mapper
@NullMarked
interface Mapper {
  Target fromSource(Source source);
}
```

**Comment by filiphr** [comment]
I finally found the time to work on this. I've created PR #4033 that is tackling this. I've decided to only support JSpecify, since it seems that the ecosystem has converged on that. Would really appreciate if all the interested parties have a look at the PR and provide any feedback that they have.

---

To answer some of the comments that people have posted.

Thanks @nealeu for sharing those links. It looks interesting from Java point of view. I feel confident that once that lands in Java we could adjust the approach I've taken in the PR to support that. We might need a bit more refactoring for treating the types differently, but it'll work. The example you provided will lead to a compiler error, unless you explicitly provide a `defaultValue` or `defaultExpression` for mapping the `lastName`.

e.g.

```java
@Mapper
@NullMarked
interface Mapper {
  Target fromSource(Source source);
}
```

leads to compiler error

>Can't map potentially nullable source property "lastName" to `@NonNull` constructor parameter "lastName". Consider adding a defaultValue or defaultExpression.

Whereas

```java
@Mapper
@NullMarked
interface Mapper {

  @Mapping(target = "lastName", defaultValue = "unknown")
  Target fromSource(Source source);
}
```
would compile.

**Comment by filiphr** [comment]
For everyone's information. The decision has ben made that MapStruct will only support JSpecify annotations.

If you are using libraries that do not use JSpecify then I would advise you to raise issues / PRs for those libraries to add support for JSpecify as well. If you can't raise PRs, then please share them with the community, maybe someone else will have the time and do contributions. Let's help and move the community forward.
